addons/binding/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/discovery/OpenSprinklerDiscoveryService.java

```diff
@@ -0,0 +1,151 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.opensprinkler.discovery;
+
+import static org.openhab.binding.opensprinkler.OpenSprinklerBindingConstants.*;
+import static org.openhab.binding.opensprinkler.internal.api.OpenSprinklerApiConstants.*;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InterfaceAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.net.util.SubnetUtils;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.opensprinkler.config.OpenSprinklerConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link OpenSprinklerDiscoveryService} class allow manual discovery of
+ * OpenSprinkler devices for OpenHab.
+ *
+ * @author Chris Graham - Initial contribution
+ */
+public class OpenSprinklerDiscoveryService extends AbstractDiscoveryService {
+    private final Logger logger = LoggerFactory.getLogger(OpenSprinklerDiscoveryService.class);
+    private final static Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashSet<ThingTypeUID>(
+            Arrays.asList(OPENSPRINKLER_THING, OPENSPRINKLERPI_THING));
+
+    private ExecutorService discoverySearchPool;
+
+    public OpenSprinklerDiscoveryService() {
+        super(SUPPORTED_THING_TYPES_UIDS, DISCOVERY_DEFAULT_TIMEOUT_RATE, DISCOVERY_DEFAULT_AUTO_DISCOVER);
+    }
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypes() {
+        return SUPPORTED_THING_TYPES_UIDS;
+    }
+
+    @Override
+    protected void startScan() {
+        logger.debug("Starting discovery of OpenSprinkler devices.");
+
+        try {
+            List<String> ipList = getIpAddressScanList();
+
+            discoverySearchPool = Executors.newFixedThreadPool(DISCOVERY_THREAD_POOL_SIZE);
+
+            for (String ip : ipList) {
+                discoverySearchPool.execute(new OpenSprinklerDiscoveryThread(this, ip));
+            }
+
+            discoverySearchPool.shutdown();
+
+            try {
+                discoverySearchPool.awaitTermination(DISCOVERY_THREAD_POOL_SHUTDOWN_WAIT_TIME_SECONDS,
```

Comment:
you must not block the startScan() method, it has to return quickly. I also do not see a reason why you should block here.